File: lib/data/services/auth_service.dart
import '../../core/network/backend_api_client.dart';
import '../models/app_user.dart';

class AuthService {
  final _client = BackendApiClient.instance;

  Future<AuthResult> signup({
    required String name,
    required String email,
    required String password,
    required String confirmPassword,
  }) async {
    final data = await _client.post(
      "/api/auth/signup",
      auth: false,
      body: {
        "name": name,
        "email": email,
        "password": password,
        "confirm_password": confirmPassword,
      },
    );
    return AuthResult.fromJson(data);
  }

  Future<AuthResult> login({required String email, required String password}) async {
    final data = await _client.post(
      "/api/auth/login",
      auth: false,
      body: {"email": email, "password": password},
    );
    return AuthResult.fromJson(data);
  }

  Future<void> logout() => _client.post("/api/auth/logout");

  Future<AppUser> me() async {
    final data = await _client.get("/api/auth/me");
    return AppUser.fromJson(data);
  }

  Future<void> forgotPassword(String email) =>
      _client.post("/api/auth/forgot-password", auth: false, body: {"email": email});
}

class AuthResult {
  final String accessToken;
  final String refreshToken;
  final AppUser user;

  AuthResult({required this.accessToken, required this.refreshToken, required this.user});

  factory AuthResult.fromJson(Map<String, dynamic> json) => AuthResult(
        accessToken: json["access_token"] as String,
        refreshToken: json["refresh_token"] as String,
        user: AppUser.fromJson(json["user"] as Map<String, dynamic>),
      );
}

File: frontend/src/services/authService.ts
import { backendApi } from "./backendApi";
import { StorageKeys, readStorage, removeStorage, writeStorage } from "@/utils/storage";
import { type AppUser, parseAppUser } from "@/types/models";

interface AuthResponse {
  access_token: string;
  refresh_token: string;
  user: any;
}

function persist(result: AuthResponse) {
  writeStorage(StorageKeys.accessToken, result.access_token);
  writeStorage(StorageKeys.refreshToken, result.refresh_token);
}

export const authService = {
  async signup(name: string, email: string, password: string, confirmPassword: string): Promise<AppUser> {
    const { data } = await backendApi.post<AuthResponse>("/api/auth/signup", {
      name,
      email,
      password,
      confirm_password: confirmPassword,
    });
    persist(data);
    return parseAppUser(data.user);
  },

  async login(email: string, password: string): Promise<AppUser> {
    const { data } = await backendApi.post<AuthResponse>("/api/auth/login", { email, password });
    persist(data);
    return parseAppUser(data.user);
  },

  async logout(): Promise<void> {
    try {
      await backendApi.post("/api/auth/logout");
    } catch {
      // Clear the local session even if the server call fails (e.g. offline).
    }
    removeStorage(StorageKeys.accessToken);
    removeStorage(StorageKeys.refreshToken);
  },

  async me(): Promise<AppUser> {
    const { data } = await backendApi.get("/api/auth/me");
    return parseAppUser(data);
  },

  /** Validates the stored token against the server rather than trusting local state. */
  async currentSession(): Promise<AppUser | null> {
    const token = readStorage(StorageKeys.accessToken);
    if (!token) return null;
    try {
      return await authService.me();
    } catch {
      return null;
    }
  },

  async forgotPassword(email: string): Promise<void> {
    await backendApi.post("/api/auth/forgot-password", { email });
  },
};
